internal/file/file.go

```diff
@@ -13,9 +13,96 @@
 
 package file
 
-import "regexp"
+import (
+	"errors"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// ErrNotRegularFile is returned when the file is not an regular file.
+var ErrNotRegularFile = errors.New("not regular file")
+
+// ErrNotDirectory is returned when the path is not a directory.
+var ErrNotDirectory = errors.New("not directory")
 
 // IsValidFileName checks if a file name is cross-platform compatible
 func IsValidFileName(fileName string) bool {
 	return regexp.MustCompile(`^[a-zA-Z0-9_.-]+$`).MatchString(fileName)
 }
+
+// CopyToDir copies the src file to dst dir. Existing file will be overwritten.
+// src file permission is preserved.
+func CopyToDir(src, dst string) error {
+	sourceFileStat, err := os.Stat(src)
+	if err != nil {
+		return err
+	}
+	if !sourceFileStat.Mode().IsRegular() {
+		return ErrNotRegularFile
+	}
+	source, err := os.Open(src)
+	if err != nil {
+		return err
+	}
+	defer source.Close()
+	if err := os.MkdirAll(dst, 0777); err != nil {
+		return err
+	}
+	dstFile := filepath.Join(dst, filepath.Base(src))
+	destination, err := os.Create(dstFile)
+	if err != nil {
+		return err
+	}
+	defer destination.Close()
+	err = destination.Chmod(sourceFileStat.Mode())
```

Comment:
we should caution against bad file permissions. Plugin executable should never be modifiable by any other user that installed it.

Reply:
> * Plugin executable should never be modifiable by any other user that installed it.

Preserving read/ exec bits and making it not writeable for `group` or `other`.

> * Plugin should always be executable (check for permission) else its useless

This was checked at the very beginning of the plugin install process. When we reach to the `copy` step, it is guaranteed that plugin is executable.